File: zs-leetcode-go/IntersectionofTwoLinkedLists/IntersectionofTwoLinkedLists.go
package IntersectionofTwoLinkedLists

import (
	"for-an-offer/zs-leetcode-go/types"
)

func getIntersectionNode(headA, headB *types.ListNode) *types.ListNode {
	if headA == nil || headB == nil {
		return nil
	}
	lenA, lenB := 0, 0
	for walk := headA; walk != nil; walk, lenA = walk.Next, lenA+1{}
	for walk := headB; walk != nil; walk, lenB = walk.Next, lenB+1{}
	pa, pb := headA, headB
	if lenA > lenB {
		for index := 0; index < lenA-lenB; index++ {
			pa = pa.Next
		}
	}
	if lenA < lenB {
		for index := 0; index < lenB-lenA; index++ {
			pb = pb.Next
		}
	}
	for ;pa != nil; pa, pb = pa.Next, pb.Next {
		if pa == pb {
			return pa
		}
	}
	return nil
}

File: sjq-leetcode-C++/intersection-of-two-linked-lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *getIntersectionNode(ListNode *headA, ListNode *headB) {
        if(headA==NULL or headB==NULL) return NULL;
        set<ListNode*> setA;
        while(headA){
            setA.insert(headA);
            headA = headA->next;
        }
        while(headB){
            if(setA.find(headB) != setA.end())
                break;
            headB = headB->next;
        }
        return headB;
    }
};
